tests/openssl_moretest/common_code/common_digest_tests.cpp

```diff
@@ -0,0 +1,735 @@
+        PKCS7_SIGNER_INFO* sitmp;
+        char* buffer = p->p[2];
+        sinfos = PKCS7_get_signer_info(p7);
+        if (!sinfos)
+        {
+            ret = 0;
+        }
+
+        // serialize partially the sinfos object in order to compare in checker
+        int num = sk_PKCS7_SIGNER_INFO_num(sinfos);
+        size_t total = 0;
+
+        _cpy(buffer, (char*)&num, sizeof(num));
+        total += sizeof(num);
+        for (int i = 0; i < num; i++)
+        {
+            int x;
+            sitmp = sk_PKCS7_SIGNER_INFO_value(sinfos, i);
+            x = sk_X509_ATTRIBUTE_num(sitmp->auth_attr);
+            _cpy(buffer + total, (char*)&x, sizeof(num));
+            total += sizeof(num);
+            total += serialize_ASN1_OBJECT(
+                buffer + total, sitmp->digest_alg->algorithm);
+            total += serialize_ASN1_OBJECT(
+                buffer + total, sitmp->digest_enc_alg->algorithm);
+        }
+    }
+    else if (_pkcs7_sign == info)
+    {
+        size_t len;
+        // write signature to buffer
+        out_bio = BIO_new(BIO_s_mem());
+        if (NULL == out_bio)
+            goto Exit;
+        ret = SMIME_write_PKCS7(out_bio, p7, in_bio, 0);
+        if (1 != ret)
+        {
+            ret = 0;
+            goto Exit;
+        }
+
+        BIO_get_mem_ptr(out_bio, &y);
+        len = (size_t)p->p[1];
+        _cpy(p->p[2], y->data, len);
+    }
+    else if (_pkcs7_verify == info)
+    {
+        STACK_OF(X509)* chain = NULL;
+
+        chain = sk_X509_new_null();
+        if (NULL == chain)
+            goto Exit;
+        sk_X509_push(chain, x509);
+        ret = PKCS7_verify(p7, chain, NULL, NULL, NULL, PKCS7_NOVERIFY);
+        sk_X509_pop(chain);
+        sk_X509_free(chain);
+    }
+    else
+    {
+        X509_STORE* store = X509_STORE_new();
+        if (NULL == store)
+            goto Exit;
+        X509_LOOKUP* lookup = X509_STORE_add_lookup(store, X509_LOOKUP_file());
+
+        ret = X509_STORE_add_cert(lookup->store_ctx, x509);
+        X509_STORE_free(store);
+    }
+
+Exit:
+
+    if (in_bio)
+        BIO_free(in_bio);
+    if (out_bio)
+        BIO_free(out_bio);
+    if (x509)
+        X509_free(x509);
+    if (pkey)
+        EVP_PKEY_free(pkey);
+    if (p7)
+        PKCS7_free(p7);
+    return ret;
+}
+
+static void Reset_EVP_MD_CTX(char* p)
+{
+    memset(p, 0, sizeof(EVP_MD_CTX));
+}
+
+const EVP_MD* some_EVP_sha(unsigned char c)
+{
+    c &= 0x3;
+
+    switch (c)
+    {
+        case 0:
+            return EVP_sha224();
+        case 1:
+            return EVP_sha256();
+        case 2:
+            return EVP_sha384();
+        case 3:
+            return EVP_sha512();
+        default:
+            break;
+    }
+    return NULL;
+}
+
+static int Create_EVP_MD_CTX()
+{
+    EVP_MD_CTX* ctx;
+    ctx = EVP_MD_CTX_new();
+    if (ctx)
+    {
+        EVP_MD_CTX_free(ctx);
+        return 1;
+    }
+    else
+    {
+        return 0;
+    }
+}
+
+int common_digest_tests(void* buf)
```

Comment:
Please do not introduce this IOCTL-style calling pattern. The entire design of this set of tests looks like it's introducing its own call marshalling infrastructure to bypass oeedger8r, and it seems largely like unnecessary overhead to funnel all these calls through what is effectively a custom function table.

Reply:
changed to `int common_digest_tests(openssl_api_param* p)` and update related files